cosim/black-parrot-example/ps.cpp

```diff
@@ -148,13 +153,22 @@ extern "C" void cosim_main(char *argstr) {
 
   long val;
   bsg_pr_info("ps.cpp: reading three base registers\n");
-  bsg_pr_info("ps.cpp: reset(lo)=%ld dram_init=%ld, dram_base=%lx\n",
-              zpl->axil_read(GP0_RD_CSR_RESET       + gp0_addr_base),
+  bsg_pr_info("ps.cpp: reset(lo)=%d dram_init=%d, dram_base=%x\n",
+              zpl->axil_read(GP0_RD_CSR_BITBANG + gp0_addr_base),
               zpl->axil_read(GP0_RD_CSR_DRAM_INITED + gp0_addr_base),
               val = zpl->axil_read(GP0_RD_CSR_DRAM_BASE + gp0_addr_base));
-
-  bsg_pr_info("ps.cpp: putting BP into reset\n");
-  zpl->axil_write(GP0_WR_CSR_RESET + gp0_addr_base, 0x0, mask1); // BP reset
+  // Reset the bsg tag master
+  zpl->tag->bsg_tag_reset(zpl, GP0_RD_CSR_BITBANG + gp0_addr_base);
```

Comment:
This is pretty good, but I think a better API is to build the tag on top of the zpl rather than having the zpl own tag like we're doing here. How about this?

```
tag = bsg_tag_bitbang(&zpl->gp0, GP0_RD_CSR_BITBANG + gp0_addr_base)
tag->reset_master()
tag->reset_client(node_id)
tag->packet_write_data(node_id, els, width, data_not_reset, nodeID, payload)
```

and then ZPL doesn't need any reference to tag

Reply:
Sounds good. I will move the tag out of zynq shell. However I think reset_client is a little bit "too high level". I would prefer just using bsg_tag_packet_write and bsg_tag_bit_write instead of reset_client to have fine-grained control.

Reply:
Fine by me